src/domain/customscript-identity.ts

```diff
@@ -171,35 +162,76 @@ export class CustomScriptIdentityWatchOnly extends Identity implements IdentityI
     }
   }
 
-  private getTemplate(isChange: boolean): string {
-    if (isChange && this.covenant.changeTemplate) return this.covenant.changeTemplate;
-    if (!this.covenant.template) throw new Error('template is missing');
-    return this.covenant.template;
-  }
+  getAddress(
+    isChange: boolean,
+    index: number,
+    constructorParams?: Record<string, string | number>
+  ): ExtendedTaprootAddressInterface {
+    if (!this.covenant.template) {
+      throw new Error('missing template');
+    }
+    const template = this.covenant.template;
+    const artifact = JSON.parse(template) as Artifact;
+    const publicKey = this.deriveMasterXPub(isChange, index, true);
 
-  getAddress(isChange: boolean, index: number): ExtendedTaprootAddressInterface {
-    const template = this.getTemplate(isChange);
-    const descriptorResult = evaluate(this.getContext(isChange, index), template);
-    const outputScript = descriptorResult.scriptPubKey().toString('hex');
-    if (!descriptorResult.taprootHashTree) throw new Error('taprootHashTree is missing');
+    if (!constructorParams) {
+      constructorParams = {} as Record<string, string | number>;
+    }
+    constructorParams[this.namespace] = publicKey;
+    const constructorArgs: Argument[] = artifact.constructorInputs.map(({ name, type }) => {
+      const param = constructorParams![name];
+      if (!param) {
+        throw new Error(`missing contructor arg ${name}`);
+      }
+      switch (type) {
+        case PrimitiveType.Bytes:
+        case PrimitiveType.Signature:
+        case PrimitiveType.DataSignature:
+        case PrimitiveType.PublicKey:
+        case PrimitiveType.XOnlyPublicKey: {
+          if (typeof param === 'string') {
+            return Buffer.from(param, 'hex');
+          }
+          const buf = Buffer.alloc(8);
+          buf.writeBigUInt64LE(BigInt(param));
+          return buf;
+        }
```

Comment:
Why you doing this? you are encoding a number when XOnlyPublicKey? 🤔

Reply:
All types are handled by Ionio already, in the case of 64bit numbers, at the moment the idea is to give them passed as `bytes`, so the user (outside marina) should care about doing the right encoding and simply passing to us as 0x prefixed hex string. Ionio will know hot to handle from there